aixtend/utils/file_utils.py

```diff
@@ -14,50 +14,167 @@
         download_file_path = save_dir / (str(uuid4()) + file_ext)
     r = _request_with_retry("get", download_url)
-    with open(download_file_path, 'wb') as f:
+    with open(download_file_path, "wb") as f:
         f.write(r.content)
     return download_file_path
 
 
-def _request_with_retry(method: str, url: str, **params) -> requests.Response:
+def _request_with_retry(method: Text, url: Text, **params) -> requests.Response:
     """Wrapper around requests with Session to retry in case it fails
 
     Args:
-        method (str): HTTP method, such as 'GET' or 'HEAD'.
-        url (str): The URL of the resource to fetch.
+        method (Text): HTTP method, such as 'GET' or 'HEAD'.
+        url (Text): The URL of the resource to fetch.
         **params: Params to pass to request function
 
     Returns:
         requests.Response: Response object of the request
     """
     session = requests.Session()
-    retries = Retry(total=5,
-                    backoff_factor=0.1,
-                    status_forcelist=[ 500, 502, 503, 504 ])
-    session.mount('https://', HTTPAdapter(max_retries=retries))
+    retries = Retry(total=5, backoff_factor=0.1, status_forcelist=[500, 502, 503, 504])
+    session.mount("https://", HTTPAdapter(max_retries=retries))
     response = session.request(method=method.upper(), url=url, **params)
     return response
-    
+
+
+def download_data(url_link, local_filename=None):
+    if local_filename is None:
+        local_filename = url_link.split("/")[-1]
+    with requests.get(url_link, stream=True) as r:
+        r.raise_for_status()
+        with open(local_filename, "wb") as f:
+            for chunk in r.iter_content(chunk_size=8192):
+                # If you have chunk encoded response uncomment if
+                # and set chunk_size parameter to None.
+                # if chunk:
+                f.write(chunk)
+    return local_filename
+
+
+def upload_data(
+    file_name: Union[Text, Path], content_type: Text = "text/csv", content_encoding: Optional[Text] = None, nattempts: int = 0
+):
+    """Upload files to S3 with pre-signed URLs
+
+    Args:
+        file_name (Union[Text, Path]): local path of file to be uploaded
+        content_type (Text, optional): Type of content. Defaults to "text/csv".
+        content_encoding (Text, optional): Content encoding. Defaults to None.
+        nattempts (int, optional): Number of attempts for diminish the risk of exceptions. Defaults to 0.
+
+    Returns:
+        URL: s3 path
+    """
+    try:
+        # Get pre-signed URL
+        team_key = config.TEAM_API_KEY
+        url = config.TEMPFILE_UPLOAD_URL
+
+        headers = {"Authorization": "token " + team_key}
+
+        payload = {"contentType": content_type, "originalName": file_name}
+        r = _request_with_retry("post", url, headers=headers, data=payload)
+        response = r.json()
+        path = response["key"]
+        # Upload data
+        presigned_url = response["uploadUrl"]
```

Comment:
Can you please explain in a comment within the code how you fetch this presigned_url? The logic of this TEMPFILE_UPLOAD_URL isn't too clear from reading the implementation.